---
title: "Filter somatic muations in the large scale anf fine scale datasets"
output: html_document
date: "2024-05-14"
---

### Dataset details

The variants were called on the 2018 and 2022 dataset. <br />
filters applied so far : SNPs removed when found in PON + Friends + "crap" + modified allele frequency vector + somatic point estimate script

# -----------------------------------------------------------------------------
# Step 1: point estimates with modified allele frequencies
# -----------------------------------------------------------------------------

Filter for mean individual read depth>4 , empty variants and singletons. 

```{r filter_pn, echo=TRUE}
knitr::opts_chunk$set(echo = TRUE)
rm(list= ls())

library(extrafont)

setwd("/Users/rozenn/GaTech Dropbox/CoS/BioSci/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/2_PS/data")
# upoad pntest data and depth info
pn_est <- read.table("pntest_som_PS.txt", sep = " ", header = F ) # 15925 x 227
depth <- read.table("depth_PS.txt", sep = "\t", header = T) # 15925 x 227
ids <- read.csv("ids_PS.txt", sep = "\t", header = T) # 227 x 4

# Filter individuals based on depth
par(family = "Times New Roman")
plot(colMeans(depth), pch = 16, xlab = "individual", ylab = "depth", ylim = c(0,80)) # Mean depth per sample
abline(h=4, col = "red")
keep <- which(colMeans(depth) >= 4) # 203

pnt_est_fil <- pn_est[,keep]  # 15925 x 2-3
ids_fil <- ids[as.numeric(keep), ] # 203 x 4
depth_fil <- depth[,keep] # 15925 x 203
# 
# write.table(ids_fil, "/Users/rpineau3/Dropbox (GaTech)/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/3_spatial_structure/data/ids_PS_sup4.txt",
#             col.names = T, row.names = F, sep = "\t")

# -----------------------------------------------------------------------------
# BINARIZATION
# -----------------------------------------------------------------------------
# pn_bin <- matrix(NaN, dim(pnt_est_fil)[1], dim(pnt_est_fil)[2])
# threshold <- 0.5
# 
# for (x in 1:dim(pnt_est_fil)[1]) {
#   
#   for (y in 1:dim(pnt_est_fil)[2]) {
#     
#     if (pnt_est_fil[x,y]>threshold) {pn_bin[x,y] <- 1 } # if proba of being heteroZ > threshold, label with 1
#     
#     else {pn_bin[x,y] <- 0 }
#     
#   }
# }

# write.table(pn_bin, "/Users/rpineau3/Dropbox (GaTech)/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/2_PS/data/pnbin_PS.txt", 
#             sep = "\t", row.names = F, col.names = F)

pn_bin <- read.table("/Users/rozenn/GaTech Dropbox/CoS/BioSci/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/2_PS/data/pnbin_PS.txt", 
             sep = "\t", header = F) # 15925 x 203
  
nInd <- 203 # number of individuals in the dataset
hets <- rowSums(pn_bin)/nInd

# remove singletons, meaning present in 1/nInd individuals AND
# present in zero individuals (left after filtering out some individuals, maybe)

pn_bin_fil <- pn_bin[c(rowSums(pn_bin)>1),] # 5776 x 203
hets <- (rowSums(pn_bin_fil))/nInd

par(family="Times New Roman", cex.lab=1.2, cex.axis=1.2, cex.main=1, cex.sub=1)
hist(hets, xlab = "Heterozygosity", ylab = "counts", 100, main = "heterozygosity per SNP, after removing singletons",
     sub = "Pando + subclone datasets",
     col = "#2ca25f")

# keep track of indices of the SNPs that make it through the filters
winners <- 1:c(dim(pn_est)[1]) # 15925
winners_v1 <- winners[rowSums(pn_bin)>1] #5776

```



# -----------------------------------------------------------------------------
# Step 2: Filtering for somatic mutations
# -----------------------------------------------------------------------------

```{r filter_for_som_mutations}
nSNP <-  5776

som <- pn_bin_fil[hets<0.8,] 
winners_v2 <- winners_v1[hets<0.8] 
hets <- rowSums(som)/nInd

# setwd("/Users/rpineau3/Dropbox (GaTech)/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/2_PS/figures/") 
# pdf(file=paste("hets_per_snp_PS.pdf", sep="" ), bg = "transparent", width=5, height=3, family = "Times New Roman")

# Check the histogram
par(family="Times New Roman", cex.lab=1.2, cex.axis=1.2, cex.main=1, cex.sub=1)
hist(hets, xlab = "Heterozygosity per SNP for somatic set", ylab = "counts", 50, main = "", 
     col = "#2ca25f")

#dev.off()

# Make a boolean vector to export the SNPs to work with
bool <- matrix(0, dim(pn_est)[1], 1)
bool[winners_v2] <- 1 

# write.table(bool, "/Users/rpineau3/Dropbox (GaTech)/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/2_PS/data/bool_filter_PS.txt", 
#             row.names = F, col.names = F)

```
# -----------------------------------------------------------------------------
# GLs for Beast and other analyses
# -----------------------------------------------------------------------------

```{r preparing_beast_nexus_files}
# Exporting PS GL file
# write.table(t(som), "/Users/rpineau3/Dropbox (GaTech)/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/3_spatial_structure/data/gl_PS_5082snps_203inds.txt", 
#              row.names = F, col.names = F)


# Exporting Pando GL file
som_pando <- som[,ids_fil$group == "pando"] # 2018 pando samples only, 5082 x 102
remove <- which(rowSums(som_pando) == 0) # empty variants - 643
som_pando <- som_pando[-remove,] # 4439 x 102
remove <- which(rowSums(som_pando)==1) # remove singletons - 482
som_pando <- som_pando[-remove,] 

ids_pando <- ids_fil$full_ID[ids_fil$group == "pando"] 

# write.table(t(som_pando), "/Users/rpineau3/Dropbox (GaTech)/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/3_spatial_structure/data/gl_pando_3957snps_102inds.txt", 
#              row.names = F, col.names = F)

write.table(t(som_pando_fil), "/Users/rpineau3/Dropbox (GaTech)/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/2_PS/beast/nexus/pando_4439snps_102inds.txt",
            row.names = F, col.names = F)
write.table(ids_pando, "/Users/rpineau3/Dropbox (GaTech)/BioSci-Ratcliff/Rozenn/4.PandoProject/17-analyses/2_PS/beast/nexus/pando_ids_102inds.txt",
            row.names = F, col.names = F)

```
